### PR TITLE
fix(@desktop/chat): chat - messages in 1-1 chat arrive out of order

### PR DESCRIPTION
**Pull request by saledjenic**
The following explanation is added to the message item file, since that may be convenient for devs to easier understand purpose of `timestamp` and `localTimestamp` instead of searching through the changes/log.

An explanation why we keep track of `timestamp` and `localTimestamp` here,
introduction of those two actually fixes the following issues:
- https://github.com/status-im/status-desktop/issues/6004
- https://github.com/status-im/status-desktop/issues/7058

We should always refer to `whisperTimestamp` as it is set for a message by the network
in order they are sent (that solves the issue #6004), but another issue #7058 is happening
cause `whisperTimestamp` has one second accuracy (which is a very big timeframe for messages).
That further means that all messsages sent by user A within 1000ms will be received with the
same `whisperTimestamp` value on the side of user B, in that case to differ the order of
those message we're using localy set `timestamp` on the sender side which is received unchanged
on the receiver side.
Now a question why don't we use only locally set `timestamp` may araise... the answer is...
because of issue #6004, cause it can be that users A and B send a message in almost the same
time in that case message sent by user A will be immediatelly added to the message list, while
message sent by user B will arrive like a less then a second later and in that case user A may
see user B message before or after his message and the same for user B, depends on local time
of those 2 users which is set for `timestamp` time in the moment they sent a message.

If we anyhow find a way to have here accutacy higher than 1 second, then we can go on only
with `whisperTimestamp`
https://github.com/status-im/status-go/blob/3f987cc565091327f017bfe674c08ed01e301d00/protocol/messenger.go#L3726

Fixes: #7058

### CODE REVIEW

**Review by osmaczko — APPROVED** [review]
Nasty bug... The fix looks good.

P.S. This is a good candidate for extraction to status-go, we probably want to share the same logic for messages ordering between different platforms. Instead of `localTimestamp` we could have `messageOrder` variable obtained from status-go, as follows:
--fetch more messages--
messageA `messageOrder: 3`
messageB `messageOrder: 2`
messageC `messageOrder: 1`
messageD `messageOrder: 0`

File: src/app/modules/shared_models/message_item.nim [hunk 1]
```diff
@@ -26,6 +51,7 @@ type
     gapFrom: int64
     gapTo: int64
     timestamp: int64
+    localTimestamp: int64
```
[the remark this answers is not in the document]
Comment (saledjenic) [reply src/app/modules/shared_models/message_item.nim:54]: @osmaczko I added these comments.

**Comment by saledjenic** [comment]
> Nasty bug... The fix looks good.
> 
> P.S. This is a good candidate for extraction to status-go, we probably want to share the same logic for messages ordering between different platforms. Instead of `localTimestamp` we could have `messageOrder` variable obtained from status-go, as follows: 

@osmaczko I still think that the best solution would be if we can have higher accuracy for `whisperTimestamp`, if it can be 1ms, then all would be good.

**Review by glitchminer — APPROVED** [review]
Tested and approved

**Comment by stefandunca** [comment]
@saledjenic significant finding.

Not sure I understand how using a local timestamp reliably fixes the message order across participants.

My understanding is that due to how local clocks work the local clock has no meaning to the rest of the participants other than providing the relative order of messages. However, given that we don't propagate the local clock to other participants it means that we only use it to keep the order of incoming messages. This in turn means we rely on the fact that `status-go` delivers messages in order. Is this right?

At first glance, I see that `status-go` uses a Lamport timestamp (see [here](https://github.com/status-im/status-go/blob/b5cc14f302c1795ba5d430cda938043b2a9c63a5/protocol/protobuf/chat_message.pb.go#L630) which is the usual way to sync order in a distributed environment if I remember correctly. Can't we propagate and use that directly instead of the local timestamp that encodes the order of messages in a non-intuitive way?

**Review by richard-ramos — APPROVED** [review]

**Comment by saledjenic** [comment]
@stefandunca whisper message has accuracy of 1s, but timestamp which is part of a message has accuracy of 1ms, that's why we cannot rely on whisper timestamp. I agree messages are coming in order, but read my first comment here, there is an example. The thing is when you send a message to me you set timestamp of that message as well, if I send a message to you within the same interval of 1000ms, but e.g. 100ms after you sent a message and I set timestamp for my message, now those 2 messages have different timestamp let's say the difference is 100ms, but their whisper timestamp is the same, then on my side my message even it's sent 100ms after you sent your will be displayed before your message.

**Comment by stefandunca** [comment]
> @stefandunca whisper message has accuracy of 1s, but timestamp which is part of a message has accuracy of 1ms ...

Thanks @saledjenic for the clarification. I just wanted to point out the [`clock`](https://github.com/status-im/status-go/blob/b5cc14f302c1795ba5d430cda938043b2a9c63a5/protocol/protobuf/chat_message.pb.go#L630) member which looks like it might be a better choice than the local timestamp and it is already there. You know better if it is an alternative or not.

**Comment by caybro** [comment]
Sorry for jumping in but we need some sort of monotonic, UTC timestamp, sth like: https://www.freedesktop.org/software/systemd/man/sd_bus_message_get_monotonic_usec.html

**Comment by stefandunca** [comment]
> Sorry for jumping in but we need some sort of monotonic, UTC timestamp, sth like: https://www.freedesktop.org/software/systemd/man/sd_bus_message_get_monotonic_usec.html

@caybro at a quick glance my understanding is that the APIs you mention are all local to the system and have inter-process appicability providing order for messages exchanged in the same system bus or OS.

My understanding of the ordering concerning current PR and that we are generally interested in our messaging system is related to the "total order of operations" in distributed systems, which is usually solved by using Lamport timestamps.

Hope this further clarifies the context of my concerns.

**Pull request merged**